File: Programmers/cpp/LEVEL_2_[3차]_압축.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

int findMaxStrIdx(int cur, string msg, unordered_map<string, int> dic) {
	string s = "";
	s += msg[cur];

	for (int i = cur + 1; i < msg.length(); i++) {
		if (dic.count(s + msg[i]) == 0) {
			return i - 1;
		}
		else {
			s = s + msg[i];
		}
	}

	return msg.length() - 1;
}

vector<int> solution(string msg) {
	vector<int> answer;
	unordered_map<string, int> dictionary;

	int idx = 0;
	char c = 'A';
	string s = "";
	s += (char)(c + idx);

	while (idx <= 25) {
		s = "";
		s += (char)(c + idx);

		dictionary[s] = idx + 1;
		idx++;
	}

	for (int i = 0; i < msg.length(); i++) {
		int endIdx = findMaxStrIdx(i, msg, dictionary); //가능한 문자열 마지막 인덱스 위치 구하기

		s = "";
		for (int j = i; j <= endIdx; j++) {
			s += msg[j];
		}

		i = endIdx;

		if (dictionary.count(s) != 0) {
			answer.push_back(dictionary[s]);
		}

		if (i < msg.length() - 1) {
			dictionary[s + msg[i + 1]] = ++idx;
		}
	}

	return answer;
}

File: Programmers/java/LEVEL_2_[3차]_압축.java
import java.util.*;

class Solution {
    private int findMaxStrIdx(int cur, String msg, Map<String, Integer> dic) {
        String s = String.valueOf(msg.charAt(cur));

        for (int i = cur + 1; i < msg.length(); i++) {
            if (!dic.containsKey(s + msg.charAt(i))) {
                return i - 1;
            } else {
                s = s + String.valueOf(msg.charAt(i));
            }
        }

        return msg.length() - 1;
    }

    public int[] solution(String msg) {
        int[] answer = {};
        List<Integer> tempAns = new ArrayList<>();
        Map<String, Integer> dictionary = new HashMap<>();

        int idx = 0;
        char c = 'A';
        while (idx <= 25) {
            dictionary.put(String.valueOf((char)(c + idx)), idx + 1);
            idx++;
        }

        for (int i = 0; i < msg.length(); i++) {
            int endIdx = findMaxStrIdx(i, msg, dictionary); //가능한 문자열 마지막 인덱스 위치 구하기

            StringBuilder sb = new StringBuilder();
            for (int j = i; j <= endIdx; j++) {
                sb.append(msg.charAt(j));
            }

            i = endIdx;

            String str = sb.toString();

            if (dictionary.containsKey(str)) {
                tempAns.add(dictionary.get(str));
            }

            if (i < msg.length() - 1) {
                dictionary.put(str + String.valueOf(msg.charAt(i + 1)), ++idx);
            }
        }

        answer = new int[tempAns.size()];
        int ansIdx = 0;

        for (int ans : tempAns) {
            answer[ansIdx++] = ans;
        }

        return answer;
    }
}